File: python/play.py
from world import World
from time import monotonic_ns
from os import environ
import math

class Play:
  WORLD_WIDTH = 150
  WORLD_HEIGHT = 40
  CLEAR_SCREEN = "\x1b[?2026h\x1b[H\x1b[2J"
  SHOW_SCREEN = "\x1b[?2026l"

  @classmethod
  def run(cls) -> None:
    world = World(
      Play.WORLD_WIDTH,
      Play.WORLD_HEIGHT,
    )

    minimal = environ.get("MINIMAL") == "1"

    if not minimal:
      print(world.render())

    total_tick = 0
    lowest_tick = math.inf
    total_render = 0
    lowest_render = math.inf

    while True:
      tick_start = monotonic_ns()
      world.dotick()
      tick_finish = monotonic_ns()
      tick_time = (tick_finish - tick_start)
      total_tick += tick_time
      lowest_tick = min(lowest_tick, tick_time)
      avg_tick = (total_tick / world.tick)

      render_start = monotonic_ns()
      rendered = world.render()
      render_finish = monotonic_ns()
      render_time = (render_finish - render_start)
      total_render += render_time
      lowest_render = min(lowest_render, render_time)
      avg_render = (total_render / world.tick)

      if not minimal:
        print(Play.CLEAR_SCREEN, end="")

      print(
        f"#{world.tick}"
        f" - World Tick (L: {Play._f(lowest_tick):.3f}; A: {Play._f(avg_tick):.3f})"
        f" - Rendering (L: {Play._f(lowest_render):.3f}; A: {Play._f(avg_render):.3f})"
      )

      if not minimal:
        print(rendered + Play.SHOW_SCREEN)

  @staticmethod
  def _f(value: float) -> float:
    # nanoseconds -> milliseconds
    return value / 1_000_000

Play.run()

File: go/play.go
package main

import (
  "fmt"
  "math"
  "os"
  "time"
)

const worldWidth = uint32(150)
const worldHeight = uint32(40)
const clearScreen = "\x1b[?2026h\x1b[H\x1b[2J"
const showScreen = "\x1b[?2026l"

func run() {
  world := newWorld(
    worldWidth,
    worldHeight,
  )

  minimal := os.Getenv("MINIMAL") != ""

  if !minimal {
    fmt.Print(world.render())
  }

  var totalTick float64
  var lowestTick float64 = math.MaxFloat64
  var totalRender float64
  var lowestRender float64 = math.MaxFloat64

  for {
    tickStart := time.Now()
    world.doTick()
    tickFinish := time.Now()
    tickTime := float64(tickFinish.Sub(tickStart).Nanoseconds())
    totalTick += tickTime
    lowestTick = min(lowestTick, tickTime)
    avgTick := totalTick / float64(world.tick)

    renderStart := time.Now()
    rendered := world.render()
    renderFinish := time.Now()
    renderTime := float64(renderFinish.Sub(renderStart).Nanoseconds())
    totalRender += renderTime
    lowestRender = min(lowestRender, renderTime)
    avgRender := totalRender / float64(world.tick)

    if !minimal {
      fmt.Print(clearScreen)
    }

    fmt.Printf(
      "#%d - World Tick (L: %.3f; A: %.3f) - Rendering (L: %.3f; A: %.3f)\n",
      world.tick,
      _f(lowestTick),
      _f(avgTick),
      _f(lowestRender),
      _f(avgRender),
    )

    if !minimal {
      fmt.Print(rendered + showScreen)
    }
  }
}

func _f(value float64) float64 {
  // nanoseconds -> milliseconds
  return value / 1_000_000
}

func main() {
  run()
}
